src/adapter/src/catalog/apply.rs

```diff
@@ -149,14 +149,23 @@ impl CatalogState {
     #[instrument(level = "debug")]
     fn apply_updates_inner(
         &mut self,
-        updates: Vec<StateUpdate>,
+        mut updates: Vec<StateUpdate>,
         retractions: &mut InProgressRetractions,
     ) -> Vec<BuiltinTableUpdate<&'static BuiltinTable>> {
         soft_assert_no_log!(
             updates.iter().map(|update| update.ts).all_equal(),
             "all timestamps should be equal: {updates:?}"
         );
 
+        // HACK: due to `ALTER SINK`, sinks can appear before the objects they
+        // depend upon. Fortunately, because sinks can never have dependencies
+        // and can never depend upon one another, to fix the topological sort,
+        // we can just always move sinks to the end.
+        updates.sort_by_cached_key(|u| match &u.kind {
```

Comment:
I'm not familiar with this method, but wouldn't this destroy the relative sorting of everything else in `updates`?

Reply:
Something like the following diff (which updates `sort_item_updates()`), would make me much more comfortable.

```
diff --git a/src/adapter/src/catalog/apply.rs b/src/adapter/src/catalog/apply.rs
index f8ec696394..decc626112 100644
--- a/src/adapter/src/catalog/apply.rs
+++ b/src/adapter/src/catalog/apply.rs
@@ -1571,7 +1571,13 @@ fn sort_updates_inner(updates: Vec<StateUpdate>) -> Vec<StateUpdate> {
     ) -> VecDeque<(mz_catalog::durable::Item, Timestamp, StateDiff)> {
         item_updates
             .into_iter()
-            .sorted_by_key(|(item, _ts, _diff)| item.id)
+            .sorted_by_key(|(item, _ts, _diff)| {
+                if item.create_sql.starts_with("CREATE SINK") {
+                    GlobalId::User(u64::MAX)
+                } else {
+                    item.id
+                }
+            })
             .collect()
     }
     let item_retractions = sort_item_updates(item_retractions);
```

Reply:
I pushed this, we'll see how it does in CI.

Reply:
> wouldn't this destroy the relative sorting of everything else in updates?

The sort method is stable so in theory it should leave everything in the same order. Your diff looks great though. Somehow I didn't notice the `sort_updates` function last night. 🙈